File: backend/app/schemas/platform_params.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator

from app.models.platform_params import GLOBAL_SCOPE, ParamValueType


class ParamVersionOut(BaseModel):
    id: uuid.UUID
    key: str
    scope: str
    value: str
    value_type: ParamValueType
    effective_from: datetime
    comment: str
    created_by_id: uuid.UUID | None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ParamCurrentOut(BaseModel):
    """One row of the admin screen: the value in force plus where it came from.

    `source` is what makes the screen honest — an operator has to be able to see
    that a number is a built-in default rather than something somebody chose.
    """

    key: str
    scope: str
    value: str
    value_type: ParamValueType
    group: str
    approved: bool
    note: str
    source: str  # "default" | "global" | "corridor"
    effective_from: datetime | None
    comment: str


class ParamSetIn(BaseModel):
    key: str
    value: str
    scope: str = GLOBAL_SCOPE
    comment: str = ""
    # Absent means "from now". A future stamp schedules the change.
    effective_from: datetime | None = None

    @field_validator("scope")
    @classmethod
    def _scope_shape(cls, v: str) -> str:
        v = v.strip()
        if not v:
            return GLOBAL_SCOPE
        if v == GLOBAL_SCOPE:
            return v
        if "->" not in v:
            raise ValueError("scope must be 'global' or a corridor like 'AE->US'")
        return v.upper()

    @field_validator("value")
    @classmethod
    def _value_present(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("value must not be empty")
        return v.strip()

File: frontend/src/api/platformParams.ts
import api from './client'

export type ParamValueType = 'percent' | 'decimal' | 'integer' | 'string'
export type ParamSource = 'default' | 'global' | 'corridor'

export interface ParamCurrent {
  key: string
  scope: string
  value: string
  value_type: ParamValueType
  group: string
  approved: boolean
  note: string
  source: ParamSource
  effective_from: string | null
  comment: string
}

export interface ParamVersion {
  id: string
  key: string
  scope: string
  value: string
  value_type: ParamValueType
  effective_from: string
  comment: string
  created_by_id: string | null
  created_at: string
}

export async function listParams(scope = 'global'): Promise<ParamCurrent[]> {
  const { data } = await api.get<ParamCurrent[]>('/api/admin/params', { params: { scope } })
  return data
}

export async function paramHistory(key: string, scope?: string): Promise<ParamVersion[]> {
  const { data } = await api.get<ParamVersion[]>(`/api/admin/params/${key}/history`, {
    params: scope ? { scope } : undefined,
  })
  return data
}

export async function setParam(input: {
  key: string
  value: string
  scope?: string
  comment?: string
  effective_from?: string
}): Promise<ParamVersion> {
  const { data } = await api.post<ParamVersion>('/api/admin/params', input)
  return data
}
